text/0000-slatepack.md

```diff
@@ -0,0 +1,425 @@
+
+#### Formatting
+
+- `WORD_LENGTH`: `15`
+  - Number of `SimpleBase58Check` encoded characters per word; chosen for human readability across device screen sizes
+
+- `LINE_BREAK`: `200` words
+  - Number of words of `WORD_LENGTH` to include before inserting a newline; chosen for user friendliness in terminals and messaging applications
+
+- `MESSAGE_BREAK`: `X` lines
+  - Number of lines an armored message formatted with `WORD_LENGTH` and `LINE_BREAK` can contain before it must be split into multiple message parts
+  - This parameter still needs to be selected, ideally to cover the most possible likely slate sizes
+
+- `MAX_MESSAGES`: `9`
+  - The maximum number of messages that can be used to represent a slate payload
+
+All chosen parameters except for `MAX_MESSAGES` should be changeable without breaking armored slates as whitespace and newlines are removed during decoding and in the case of message breaks, the error check code is over the entire slate payload. `MAX_MESSAGES` cannot exceed 9 because they would be rejected by clients as invalid since the regex will not match. Slates larger than this should use synchronous transport for a better user and developer experience.
+
+### Example Slatepack JSON Object
+
+#### Plain Text
+
+In this plain text example, neither the sender nor the receiver wish to share a `SlatepackAddress`
+
+```
+{
+  "slatepack": [0, 1],
+  "mode": 0,
+  "sender": "0",
+  "payload": <binary serialized slate>
+}
+```
+
+#### Encrypted
+
+```
+{
+  "slatepack": [0, 1],
+  "mode": 1,
+  "sender": "slatepack1p4fuklglxqsgg602hu4c4jl4aunu5tynyf4lkg96ezh3jefzpy6swshp5x",
+  "header": {
+    "recipients_list": {
+        "0": {
+            "epk": "SVrzdFfkPxf0LPHOUGB1gNb9E5Vr8EUDa9kxk04iQ0o",
+            "emk": "0OrTkKHpE7klNLd0k+9Uam5hkQkzMxaqKcIPRIO1sNE"
+        }
+    },
+    "mac": "gxhoSa5BciRDt8lOpYNcx4EYtKpS0CJ06F3ZwN82VaM"
+  },
+  "payload": <age encrypted binary seriaized slate binary>,
+}
+```
+
+### Example `SlatepackMessage`
+
+```
+BEGINSLATEPACK. 4H1qx1wHe668tFW yC2gfL8PPd8kSgv
+pcXQhyRkHbyKHZg GN75o7uWoT3dkib R2tj1fFGN2FoRLY
+GWmtgsneoXf7N4D uVWuyZSamPhfF1u AHRaYWvhF7jQvKx
+wNJAc7qmVm9JVcm NJLEw4k5BU7jY6S eb. ENDSLATEPACK.
+```
+
+## `SlatepackWorkflow`
+
+Adoption of the Slatepack standard allows for a unified workflow that can still function without knowledge of a `SlatepackAddress` from a counterparty.
+
+### With a `SlatepackAddress`
+
+1. `grin-wallet send -d slatepack1bech32address 1.337`
+
+2. Sender wallet derives an onion v3 address from `slatepack1bech32address` and attempts to complete the transaction synchronously via Tor
+
+3. (Fallback) If the synchronous transaction fails, a `SlatepackMessage` string is encrypted to the `SlatepackAddress` and output for manual asynchronous transport by the user
+
```

Comment:
Implementation detail, I think for the purposes of the command line wallet it might be worthwhile outputting a file as well as displaying the slatepack string to the user, particularly due to any issues that might exist cutting/pasting from the particular terminal

Reply:
Just to follow up with some reasoning here, say you're armoring/encrypting a message using GPG, the default is to output to the terminal, and you need to provide a flag to save it to file. If you get it wrong the first time and you actually wanted to output to file, no big deal, run the command again with `-o filename` or whatever it is.

However, if you run a grin-wallet command such as `init_send_tx`  and you lose your output somewhere in your terminal, you've locked a load of inputs and changed the wallet state. You can't simply re-run the command to get the SlatepackMessage again, first you'd need to cancel the previous tx, etc.

We may be able to revisit this with lock-free transactions, but for now I'm leaning toward thinking a file should be output somewhere in a `slatepack` directory, and a message given to the user that this slatepack has also been output to this location. Still allowing a --file flag as well if the user wants to output to file explicitly.

Reply:
>I think for the purposes of the command line wallet it might be worthwhile outputting a file as well as displaying the slatepack string to the user

I agree, this is can be done in the background, has a low cost and can help ensure reliability when building txs as illustrated by your example. 

>a file should be output somewhere in a slatepack directory, and a message given to the user that this slatepack has also been output to this location

This is reasonable. It is also important to notify the user that the data is lingering somewhere on their disk so they can wipe it if they desire. Maybe it would be desirable to have it as a config setting with that behavior as default.

This approach also aligns well with the proposed changing of the edge case fallback to file as opposed to multipart messages (users will be conditioned to where they can find their slatepack if tor is unavailable and the output would be too large for a copy/paste)